posts/hello-microservice-deployment-part-2.md

```diff
@@ -0,0 +1,358 @@
+
+The diagram here shows just a small portion of the Webflix application. Let's assume it's peak watching time, the little numbers next to the various services are examples of the number of live containers in play. In theory you would be able to make this happen through use of the `docker run` command.
+
+Now let's say one of the recommend_service containers hits an unrecoverable error and falls over. If we were managing this through `docker run` then we would need to notice this and then restart the offending container. Or suddenly we are no longer in peak time and the numbers of containers needs to be scaled down. Or you want to deploy a new version of the recommend_service image? Then there are other concerns like container startup time and aliveness probes. And what if you have a bunch of machines that you want to run your containers on, you would need to keep track of what resources are available on each machine, and what replicas are running on each machine so you can scale up and scale down when you need to.
+
+That is where container orchestration comes in.
+
+Please note, the author of this post is terribly biased and thinks K8s is the shizzle. So that is what we will cover here. There are alternative platforms that you might want to look into, such as Docker Swarm and Mesos/Marathon.
+
+Ok, back to the K8s goodness.
+
+K8s was born and raised and battle hardened at Google. Google used it internally to run huge production workloads, and then they open-sourced it because they are lovely people. So K8s is a freely available, highly scalable, extendable and generally solid platform. Another thing that is really useful is that Google cloud has first class K8s support, so if you have a K8s cluster on Google cloud then there are all sorts of tools available that make managing your application easier.
+
+K8s defines a bunch of concepts, there are entire books written on the subject so what follows is a surface level introduction. Personally I've found [this book](https://www.amazon.com/gp/product/1491935677/ref=as_li_tl?ie=UTF8&camp=1789&creative=9325&creativeASIN=1491935677&linkCode=as2&tag=sheena0d-20&linkId=e5bf88f134212c4beba73c601598d7ff") seriously useful.
+
+K8s is all about managing certain objects. These objects can be created and configured in a bunch of ways. We'll be using the command line utility kubectl. K8s lets you run your workload on a cluster of machines, K8s installs a daemon on each of these machines, the daemons manage the K8s objects. Object configuration is communicated to the cluster. K8s job is to try to make the object configuration represent reality. Eg if you state that there should be three instances of the container recommend_service:1.0.1 then K8s will create those containers. If one of the containers dies (even if you manually kill it yourself) then K8s will make the configuration true again by recreating the killed container somewhere on the cluster.
```

Comment:
1. Define a Daemon
2. The first couple sentences of this paragraph are difficult to understand. You could say --
Kubernetes lets you run your workload on a cluster of machines. It installs a daemon on each of these machine and the daemons manage the Kubernetes objects. Daemons are ... . K8s job is to try to make the object configuration represent reality. Eg if you state that there should be three instances of the container recommend_service:1.0.1 then K8s will create those containers. If one of the containers dies (even if you manually kill it yourself) then K8s will make the configuration true again by recreating the killed container somewhere on the cluster.

I think this conveys the same point.

Reply:
done